design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,656 @@
+  external resource]'s `.network` and `.subnetwork` fields.
+* Add support for `GKECluster` managed resources and their resource classes to
+  specify the [GKE cluster external resource]'s `.network` field and
+  `.subnetwork` fields, and specify the external names of a `Network` and a
+  `Subnetwork` managed resource.
+* Use the `GKECluster` managed resource's existing support for setting the
+  [GKE cluster external resource]'s `.ipAllocationPolicy.useIpAliases` field.
+
+High-fidelity Crossplane managed resource representations of the aforementioned
+network and subnetwork external resources would look as follows:
+
+```yaml
+---
+apiVersion: vpc.gcp.crossplane.io/v1alpha1
+kind: Network
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  nameFormat: mycoolname
+  description: A really cool VPC network
+  # IPv4Range puts the network in subnetwork-less 'legacy mode'
+  IPv4Range: 10.0.0.0/8
+  # The documentation says this puts the VPC network in auto-create subnet mode.
+  # In my experiments it seemed to create the VPC in legacy mode instead.
+  autoCreateSubnetworks: true
+  peerings:
+  - name: peerwithsomeothercoolnetwork
+    network: someothercoolnetwork
+    autoCreateRoutes: true
+    exchangeSubnetRoutes: true
+  routingConfig:
+    routingMode: REGIONAL
+```
+
+```yaml
+---
+apiVersion: vpc.gcp.crossplane.io/v1alpha1
+kind: Subnetwork
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  nameFormat: mycoolname
+  description: My cool VPC subnetwork
+  network: projects/coolproject/global/networks/coolestvpc
+  ipCidrRange: 192.168.0.0/24
+  region: us-central-1
+  privateIpGoogleAccess: true
+  secondaryIpRanges:
+  - rangeName: pods
+    ipCidrRange: 10.0.0.0/8
+  - rangeName: services
+    ipCidrRange: 172.16.0.0/16
+  enableFlowLogs: true
+  logConfig:
+    enable: true
+    flowSampling: 0.5
+    metadata: INCLUDE_ALL_METADATA
+    # The documentation helpfully says this is an enum but does not document the
+    # possible values. ¯\_(ツ)_/¯
+    aggregationInterval: FREQUENTLY
+```
+
+Putting this all together, the infrastructure administrator would configure the
+following to ensure that when an app operator created a `MySQLInstance` claim
+and `KubernetesCluster` claim the two would have connectivity:
+
+```yaml
+---
+# A Network managed resource.
+apiVersion: vpc.gcp.crossplane.io/v1alpha1
+kind: Network
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  providerRef:
+    namespace: crossplane-system
+    name: example
+  nameFormat: mycoolnetwork
+  autoCreateSubnetworks: false
+---
+# A Subnetwork managed resource.
+apiVersion: vpc.gcp.crossplane.io/v1alpha1
+kind: Subnetwork
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  providerRef:
+    namespace: crossplane-system
+    name: example
+  nameFormat: mycoolsubnetwork
+  # Create this subnet in the Network we created previously.
```

Comment:
Is there a way for this to be derived from something less specific? Or, how will users generate this value?

Reply:
Can you elaborate a little on what you're asking? I don't follow.

Reply:
i think @suskin is looking for an easier way for users to get this value correct.  could we simplify this by only requiring the user to enter `mycoolnetwork`, then when we make the GCP API call we can generate this fully qualified name, potentially re-using the project value from the provider ref this object already has set above.

Reply:
> Could we simplify this.

We potentially could, but I don't feel we should. At least not at the managed resource level. I believe Crossplane managed resources (as defined in this doc) should be high fidelity, faithful, and unsurprising translations of the underlying cloud provider APIs. This is useful for a few reasons:

* It makes our API predictable to cloud administrators. The `network` Crossplane field means exactly what the `network` field means in the upstream cloud provider API. The valid values are exactly what is valid in the upstream cloud provider API. There is no additional layer of cognitive burden in which you must translate what the cloud provider expects to what Crossplane expects and vice versa.
* It keeps the path clear for automatically generated managed resource logic.  I expect supporting "friendlier"/"easier" inputs would make our code generation more complicated than if we simply generated a client that took the managed resource authors's input and provided exactly that to the cloud API.

Do you think everyone is on the same page about this? Some of the discussions and commentary I've seen around managed resources, as well as some of the current implementations, make me think we're not entirely aligned as to whether managed resources should be faithful translations of cloud provider APIs, or whether they should provide "value add", such as making inputs simpler, provisioning IAM roles, etc.

Reply:
To be clear, I do think providing this kind of value add is in scope for Crossplane as a project, but I believe it should be built at a higher level than managed resources.

Reply:
Yes, what I was getting at was pretty much what @jbw976 said, which is it seems like as a user I would want to use a reference to the network I was already specifying.

> I believe Crossplane managed resources (as defined in this doc) should be high fidelity

It seems as though the resource types are using generic names (`Network`, `Subnetwork`) but provider-specific parameters (`privateIpGoogleAccess`, the format of the network name). If this document is proposing high-fidelity resources, would it make sense to use different names for the `Network` and `Subnetwork` Kinds? I am wondering how we will reconcile having high-fidelity objects for multiple providers with the generalized name and the Google-specific parameters.

Reply:
> It seems as though the resource types are using generic names

They are of `kind: Network` and `kind: Subnetwork` because this is what Google calls them. Note that they are in the `vpc.gcp.crossplane.io` API group. So they're really `network.vpc.gcp.crossplane.io/v1alpha1`, etc.